doc/compiler/osr/OSR.md

```diff
@@ -0,0 +1,275 @@
+These include calls, asyncchecks and monitor enters. There is also an implicit OSR yield point in the method prologue due to the stack overflow check.
+NextGenHCR is implemented using post-execution OSR, such that OSR happens after an OSR yield point is executed.
+
+During IL generation, **potentialOSRPointHelper** is inserted immediately after every OSR yield point.
+
+During inlining, the traditional HCR implementation runs, wrapping calls in HCR guards.
+
+The **OSRGuardInsertion** optimization is performed after inlining and determines when an HCR guard can be replaced with a set of OSR guards based on fear analysis. The optimization uses the following concepts:
+
+**Fear point**: a point in the IL where the JIT has transformed the code in a way that depends on one or more assumptions that can be violated at an OSR yield point. As an example, a fear point can be the entry point of an inlined method that can potentially be redefined. Another example of a fear point is a point at which a static final field load has been constant-folded (see [Static Final Field Folding](#static-final-field-folding)).
+
+**HCR guard analysis**: a forward union analysis that determines at each program point whether the most recently run OSR yield point may have been uninducible. This analysis determines where fear points can be introduced, and therefore where assumptions can be made. A fear point can be introduced iff the most recent OSR yield point is necessarily inducible. Recall that method entry always contains an inducible OSR yield point. In particular, this analysis is used to determine which HCR guards may be treated as fear points and subsequently eliminated.
+
+**Fear point analysis**: is a backward union analysis that determines whether it's possible for execution to reach a fear point without first hitting an OSR yield point. OSR guards are inserted after each OSR yield point at which this is possible. This ensures that every fear point is protected by the closest possible OSR guard. In addition, all of the resulting OSR guards are in positions where it's possible to induce OSR because HCR guard analysis has prevented the creation of any fear points that would require OSR guards where OSR is uninducible.
+
+OSR guards are placed after OSR yield points and behave similarly to HCR guards. The guard will not necessarily be placed immediately after the OSR yield point,
+rather several stores may be located between the two. These will include stores to pending push slots, to determine the stack at the transition point (as these are used for rematerialization),
+and stores of call or allocation results. The fall through path will be a path to a fear point without going through any other OSR guards,
+but the taken path will point to an **OSR induce block**. OSR induce block will contain call to the **jitInduceOSRAtCurrentPC** helper and will have exception edges to the **OSR catch block**.
```

Comment:
*The* OSR induce block will contain *a* call to...

Reply:
Still need to add *a*: will contain *a* call to

Reply:
done